gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import static com.google.cloud.pubsub.spi.v1.SubscriberApi.formatSubscriptionName;
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+import com.google.cloud.GrpcServiceOptions.ExecutorFactory;
+import com.google.cloud.pubsub.PubSub.MessageConsumer;
+import com.google.cloud.pubsub.PubSub.MessageProcessor;
+import com.google.cloud.pubsub.spi.PubSubRpc;
+import com.google.cloud.pubsub.spi.PubSubRpc.PullCallback;
+import com.google.cloud.pubsub.spi.PubSubRpc.PullFuture;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+
+import io.grpc.internal.SharedResourceHolder;
+
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Default implementation for a message consumer.
+ */
+final class MessageConsumerImpl implements MessageConsumer {
+
+  private static final int MAX_QUEUED_CALLBACKS = 100;
+  // shared scheduled executor, used to schedule pulls
+  private static final SharedResourceHolder.Resource<ScheduledExecutorService> TIMER =
```

Comment:
Maybe fine for now, however I think we need to co-ordinate efforts with grpc-java or VTKIT to have shared resources or at least shared code to initialize the resources such that they will work on every platform. Currently the default TIMER and EXECUTOR will not work in AE standard.